### PR TITLE
Only iterate the files that we recovered from the commit

### PR DESCRIPTION
**Pull request by s1monw**
Today we use Directory#listAll() to find all the files we recovered. Yet,
this is not accurate since there might be leftovers etc. It's better to
only iterate over the known files from the segments info that we recovered.

### CODE REVIEW

**Comment by kimchy** [comment]
LGTM

**Comment by bleskes** [comment]
LGTM do we want a test that injects evil files?

**Comment by s1monw** [comment]
> LGTM do we want a test that injects evil files?

that one is going be available at the lucene release next to you soon

File: src/main/java/org/elasticsearch/common/lucene/Lucene.java [hunk 1]
```diff
@@ -121,6 +119,17 @@ public static SegmentInfos readSegmentInfos(Directory directory) throws IOExcept
     }
 
     /**
+     * Returns an iterable that allows to iterate over all files in this segments info
+     */
+    public static Iterable<String> files(SegmentInfos infos) throws IOException {
+        final List<Collection<String>> list = new ArrayList<>();
+        for (SegmentCommitInfo info : infos) {
```
Comment (rmuir) [inline src/main/java/org/elasticsearch/common/lucene/Lucene.java]: Should this include the segments file? (e.g. call infos.files() instead) The old listAll() would have included it too.

Comment (s1monw) [inline src/main/java/org/elasticsearch/common/lucene/Lucene.java]: pushed a fix for this here https://github.com/elasticsearch/elasticsearch/commit/e8f276b261806eff933a56015b6c29d38dc876d8

**Pull request merged**